### ISSUE TITLE
Proposal: Modify Extended Headers

### ISSUE THREAD

**Issue by MarkDavidson** [TAXIIProject/TAXII-Specifications#65]
In TAXII 1.1 XML Message Binding, Extended Headers are implemented as an `xs:any`, which makes them hard to translate across technologies (e.g., JSON). This issue proposes a simplification: Attempt to limit Extended Headers to simple key/value pairs where values can only be scalars (e.g., string, int, boolean). Values would be prohibited from being objects, which are currently  permitted in TAXII 1.1.

This investigation will require identifying use cases for extended headers and attempting to identify use cases that can not be met with scalar values.

-Mark

**Comment by MarkDavidson** [comment]
Per Dave Cridland on the TAXII Discussion List:

> 4) Extension data.

> As I alluded to in another message, one of the benefits of XML is that it has namespacing built-in, which allows for some fairly useful constructs. XMPP uses this heavily, as one might expect from a XML-native protocol, but I suspect it would be possible to use something fairly similar in §4.1.5 and similar places. I'd go so far as to say that extended headers, for example, would be best defined as a tuple of namespace, local name, and value, as with XML Namespaces, since that will map to both XML and JSON neatly - but be significantly more compact in XML.

**Comment by jordan2175** [comment]
My biggest complaint about the whole "it has namespaces" aspect is, that unless both parties have implemented it in code, then having namespaces does not really matter.  Just because I can point to a namespace to find the schema does not mean that my code will automatically create structs, classes, and objects, to handle the data.  

I would really like to see the use cases for the extended_headers, or as I would like to see them called in the future "options", and see if there is any valid use case where something other than key/value pairs are needed.  

We also need to keep in mind that if we build something to complex so as to accommodate every possible corner case out there, then the chances of that feature actually being implemented in code goes down significantly.  People will just ignore it or worse, make an assumption that they will only have a limited set of the data.

Can we get a list of valid use cases attached to this issue?

**Comment by MarkDavidson** [comment]
@jordan2175,

There aren't any concrete use cases for this feature that I'm currently aware of. We may have to build them in the OASIS working group. 

I wholeheartedly agree that future solutions should be of minimal complexity to support defined use cases. I also agree that adding complexity for corner cases is not ideal. 

If people have use cases for this feature, then please attach them 

Thank you.
-Mark

**Comment by stkyle** [comment]
Can someone elaborate on this issue?  I like having a relatively permissive block in TAXII messages because it gives service providers the flexibility to include arbitrarily robust 'add-ons' without breaking the specs.
Thanks.
Steve

**Comment by MarkDavidson** [comment]
@stkyle,

From my perspective, this is due to the complexities of supporting `xs:any` in the following two scenarios:

1. In a general sense (e.g., in libtaxii)
2. Across various technologies (e.g., XML->JSON)

I'll attempt to start my discussion from the reverse direction since I feel that may make more sense.

In any protocol, there's value in having extension points; to my knowledge, most protocols have at least one (HTTP/SMTP X-Headers, etc). My current opinion - subject to change due to well reasoned arguments - is that key/value pairs can satisfy the large majority of, and possibly all, extension needs. It seems to me that complex extension needs can be met by using multiple key/value pairs or pushing complexity into the value portion of the key/value pair. For instance, you could put a whole XML document in there if you wanted.

I am very interested in arguments against my previous paragraph, since that it the basis for my current reasoning.

I then ask: What is the added cost, and what is the added benefit, of allowing something more complex than a key/value pair? For me right now, the cost is apparent and the benefit is not. The costs are:

* A general piece of software (like libtaxii) should support infinitely deep, infinitely complex values since these are allowed as commonplace (this is the current case - you will notice that libtaxii draws a line in the sand and does not have full support). In they key/value case, you can simply use native language constructs (e.g., a dictionary in Python) to fully implement the feature. In the `xs:any` case, you pretty much need a generalized XML parser.
* Within an XML format, you can get numbers, strings, datetimes, XML, or many other kinds of data. There should be logic to have schema-aware parsing of the XML so that the XML can be deserialized into a native python object (when possible). With a key/value format, it's just strings, and it's up to the application using the general library to make sense of things.

In short, it seems to me that the "simple" key/value case can handle everything that the key/xs:any case can, but is simpler to implement and is more portable across languages (Java, Python) and formats (XML, JSON). A key counter-argument here would be to find an important use case that the "simple" key/value format can not fulfill that the key/xs:any format can.

Thank you.
-Mark

**Comment by terrymacdonald** [comment]
Thanks Mark,

You make a good case in my opinion.

Cheers

*Terry MacDonald *| STIX, TAXII, CybOX Consultant

M: +61-407-203-026
E: terry.macdonald@threatloop.com
W: www.threatloop.com

<https://www.threatloop.com>

*Disclaimer: The opinions expressed within this email do not represent the
sentiment of any other party except my own. My views do not necessarily
reflect those of my employers.*
